### PR TITLE
ABC-168 Update initial scrolling on post list

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
This should hopefully fix the scrolling/flicker issues when switching to a channel. I'm going to have to ask for QA's and/or the PMs' help with testing this to check all the corner cases. These are the possibilities to consider, switching to a channel:

* You have zero posts for and has no unreads
* You have zero posts for and has unreads (two cases here, scrolled to bottom or scrolled off bottom to new message indicator)
* You have some posts for but you haven't loaded the channel before (same two cases as above)
* You have some posts for but missing some newer messages from a disconnected WS (same two cases as above)
* You have loaded before and has no unreads
* You have loaded before and has unreads that you have from the WS (same two cases as above)
* You have loaded before and has unreads that you're missing from a disconnected WS (same two cases as above)
* All the above cases except with a slow connection

I've included the ability to manually disconnect the websocket by opening the dev console and entering `close();` to help with testing. To reconnect, just refresh the page.
screen shot 2018-01-25 at 9 46 01 am

I've tried all these cases myself and it seems to work well but I might have missed something.

#### Ticket Link
https://mattermost.atlassian.net/browse/ABC-168

### CODE REVIEW

**Comment by lindy65** [comment]
@jasonblais I've tested - results below...

* You have zero posts for and has no unreads
   * no issues

* You have zero posts for and has unreads (two cases here, scrolled to bottom or scrolled off bottom to new message indicator)
   * no issues

* You have some posts for but you haven't loaded the channel before (same two cases as above)
   * beginning of channel flashes for a split second before loading the posts (if I join the channel from the "+" rather than being added to it by another user)

* You have some posts for but missing some newer messages from a disconnected WS (same two cases as above)
   * no issues for either case

* You have loaded before and has no unreads
   * no issues

* You have loaded before and has unreads that you have from the WS (same two cases as above)
   * no issues for either case

* You have loaded before and has unreads that you're missing from a disconnected WS (same two cases as above)
   * no issues for either case

* All the above cases except with a slow connection
   * no issues for all tests except:
      * You have some posts for but you haven't loaded the channel before (same two cases as above)
         * where I see the same behavior as above:
            * beginning of channel flashes for a split second before loading the posts (if I join the channel from the "+" rather than being added to it by another user)

**Comment by jasonblais** [comment]
@jwilander Can you take a look at this comment?

> You have some posts for but you haven't loaded the channel before (same two cases as above)

> beginning of channel flashes for a split second before loading the posts (if I join the channel from the "+" rather than being added to it by another user)

Linda will also help with a round of testing, on Tuesday/Wednesday.

**Comment by jwilander** [comment]
Yeah, I'll take a look. Just confirming, but by beginning of channel we mean this bit right?
screen shot 2018-01-26 at 1 46 53 pm

**Comment by lindy65** [comment]
Hey @jwilander - yes, as per your screenshot...

**Comment by jwilander** [comment]
Ok, the flickering of the beginning of the channel is from showing the "Load more messages" immediately when the channel is first loaded and then when the network request to get the rest of the posts from the server completes it either knows if it has more posts or if we're got them all. If it has more posts, it just loads those in. If it it doesn't have more posts, we know we can show the beginning of the channel.

I'm proposing we don't fix for this since it should only happen in channels with little enough posts that the beginning of channel is shown when switching to it. What might be worth it, is to replace the initial "Load more messages" with a loading indicator until that first network request is done

**Comment by jasonblais** [comment]
@jwilander How much work is it to add a loading indicator?

While it wouldn't have much of an impact on an established server like pre-release, those trying out Mattermost would see it more often and potentially worsen the first impression.

We initially had a loading indicator here: https://github.com/mattermost/mattermost-server/pull/6501

If it's non-trivial, we can consider creating a separate ticket since it's not a regression from existing behaviour.

**Comment by jasonblais** [comment]
@lindalumitchell Just a reminder to have your help test this PR.

PS: I gave it a smoke test and it's working great.

**Comment by jwilander** [comment]
There's still a loading indicator for the base case like in that PR you linked, that hasn't changed with this PR. The only time the flickering would happen is when you receive posts down the WS before you've loaded that channel previously. Then the beginning of channel message might appear briefly until the network request completes.

Adding a loading indicator for that case would be about 2 mana. 1/5 that it belongs in another ticket, but not a huge deal to add here

**Comment by jasonblais** [comment]
@jwilander If it's not too much work, let's add the loading indicator. @lindalumitchell will probably test this PR once anyways before we put it to dev review.

**Comment by jwilander** [comment]
Alright I took a quick stab at it but it didn't really help the flickering, just replaced the momentary appearance of "Load more messages" with a momentary appearance of "Loading...". For slow connections it probably helps but for fast connections it didn't make a difference and the flicker happened anyways.

I'll still push the changes but getting rid of that flicker will be a bit more work and involve adding some store data to track so I think that's better for an improvement ticket

**Comment by jasonblais** [comment]
@jwilander Just needs a rebase and I'll set up a test server for Linda

**Comment by jwilander** [comment]
Done

**Comment by lindalumitchell** [comment]
Thanks guys, will test this one now.

File: actions/websocket_actions.jsx [hunk 1]
```diff
@@ -72,6 +72,7 @@ export function initialize() {
     WebSocketClient.initialize(connUrl);
 }
 
+window.close = close;
```
Comment (enahum) [inline actions/websocket_actions.jsx]: is this intentional?

Comment (jwilander) [reply actions/websocket_actions.jsx]: Did you read the summary? :P

>I've included the ability to manually disconnect the websocket by opening the dev console and entering close(); to help with testing. To reconnect, just refresh the page.

I'll remove it before dev review

**Comment by lindalumitchell** [comment]
@jwilander @jasonblais I ran through the tests, and it looks pretty great! I did notice that the beginning of a channel on a slow connection shows the Loading... indicator, and looks good to me.

One question, is this expected: If I'm offline when another user adds me to a channel, then I reconnect and click around through channels or post some messages to verify I'm connected, I have to refresh or switch teams and back to see the channels I was added to appear in the LHS (and the mentions on those channels).

Other than that, no new issues found!

**Comment by jasonblais** [comment]
@lindalumitchell I don't think that's expected - or at least something we could do better. Does it repro on `master`?

@jwilander Would that be something we can create a ticket for, assuming it's not a regression? Not sure how hard of a change it would be.

**Comment by lindalumitchell** [comment]
Oops, @jwilander, I just noticed you pointing out to EN that the part about refreshing the page is actually in the summary.  :)  

@jasonblais I also did just confirm that master requires a refresh to show the new channel as well. So I could put in a new ticket, depending on JW's answer to your question.

**Comment by jwilander** [comment]
Yeah that's definitely something we can fix, should be easy enough (does belong in a different ticket/PR though)

**Comment by jwilander** [comment]
If there's no objections I think we're good to get this dev reviewed, I removed the temporary commit that was for testing

**Comment by lindalumitchell** [comment]
No objections here. And I created an improvement ticket for the refresh issue: https://mattermost.atlassian.net/browse/PLT-8671

**Pull request merged**